### PR TITLE
[11930] Adds support for multiple backload

### PR DESCRIPTION
**Pull request by kenvac**
Story: 11930

### CODE REVIEW

**Review by apatrick88 — APPROVED** [review]

File: addons/udes_delivery_control/controllers/main.py [hunk 1]
```diff
@@ -19,12 +19,21 @@ def _get_picking_info(self, pickings, fields_to_fetch=None):
...
+            if "u_backload_ids" in fields_to_fetch:
+                picking_dict["u_backload_ids"] = [
+                    {"id": x.id, "supplier_name": x.supplier_id.name}
+                    for x in picking.u_backload_ids
+                ]
+            vals.append(picking_dict)
+        return vals
 
-    @http.route("/api/delivery-control/stock-picking/", type="json", methods=["GET"], auth="user")
+    @http.route("/api/delivery-control/", type="json", methods=["GET"], auth="user")
```
Comment (ws-mps) [inline addons/udes_delivery_control/controllers/main.py:36]: Can we also rename the class please? `PickingApi`

File: addons/udes_delivery_control/controllers/main.py [hunk 2]
```diff
@@ -19,12 +19,21 @@ def _get_picking_info(self, pickings, fields_to_fetch=None):
...
     def get_pickings_delivery_control(self, fields_to_fetch=None, **kwargs):
```
Comment (ws-mps) [inline addons/udes_delivery_control/controllers/main.py:37]: I would also rename the function name.

File: addons/udes_delivery_control/controllers/main.py [hunk 3]
```diff
@@ -37,7 +46,7 @@ def get_pickings_delivery_control(self, fields_to_fetch=None, **kwargs):
         return self._get_picking_info(pickings, fields_to_fetch)
 
     @http.route(
-        "/api/delivery-control/stock-picking/<int:ident>",
+        "/api/delivery-control/<int:ident>",
```
Comment (ws-mps) [inline addons/udes_delivery_control/controllers/main.py:49]: I would also rename the function name.

File: addons/udes_delivery_control/models/stock_picking.py [hunk 4]
```diff
@@ -284,6 +246,14 @@ def write(self, values):
 
     @api.model
     def create(self, values):
+        picking_type_dc = self.env.ref("udes_delivery_control.picking_type_delivery_control")
+        picking_type_id = values.get("picking_type_id")
+        if picking_type_id and picking_type_id == picking_type_dc.id:
+            # Set loading type to unload if picking is of delivery control type
+            # Note: u_loading_type is of related type field and compute attribute
+            # is not compatible/called. Hence, we are setting it's default value in create method
+            values["u_loading_type"] = "unload"
```
Comment (ws-mps) [inline addons/udes_delivery_control/models/stock_picking.py:255]: Could we just do this as default value for DC ?

File: addons/udes_delivery_control/controllers/main.py (hunk 1 above)
Comment (kenvac) [reply addons/udes_delivery_control/controllers/main.py:36]: How does `DeliveryControlApi` sound?

File: addons/udes_delivery_control/models/stock_picking.py (hunk 4 above)
Comment (kenvac) [reply addons/udes_delivery_control/models/stock_picking.py:259]: I tried all approaches but none seems to work
1. set default on fields - Using this method we have no way to get the value of picking type unless we set the context from where we create picking. 
2. Override default_get method - When the values are set in this method it gets changed by odoo somewhere as the field u_loading_type is a related field
3. compute on-field - Does not get called at all. Might be something to do with a related type of field

File: addons/udes_delivery_control/controllers/main.py (hunk 3 above)
Comment (kenvac) [reply addons/udes_delivery_control/controllers/main.py]: done

File: addons/udes_delivery_control/controllers/main.py (hunk 2 above)
Comment (kenvac) [reply addons/udes_delivery_control/controllers/main.py]: done

File: addons/udes_delivery_control/models/picking_extra_fields.py [hunk 5]
```diff
@@ -134,10 +95,11 @@ def _check_unloading_dates(self):
...
 
-    @api.constrains("backload_start_date", "backload_end_date")
-    def _check_backload_dates(self):
-        msg = _("Backloading start time cannot be greater than end time")
-        validate_dates(self.backload_start_date, self.backload_end_date, msg)
+    @api.constrains("loading_type", "backload_ids")
+    def _check_backload_ids(self):
+        if self.is_backload and not self.backload_ids:
+            msg = _("Please add backloads.")
+            raise ValidationError((msg) % self.backload_ids)
```
Comment (ws-mps) [inline addons/udes_delivery_control/models/picking_extra_fields.py:102]: Why do we need this ` % self.backload_ids` ?

File: addons/udes_delivery_control/models/stock_picking.py (hunk 4 above)
Comment (ws-mps) [reply addons/udes_delivery_control/models/stock_picking.py:259]: Fair enough, worthed asking though :)

File: addons/udes_delivery_control/models/picking_extra_fields.py (hunk 5 above)
Comment (kenvac) [reply addons/udes_delivery_control/models/picking_extra_fields.py]: Removed

**Review by ws-mps — APPROVED** [review]

**Pull request merged**